core/src/main/scala/kafka/controller/ControllerChannelManager.scala

```diff
@@ -138,9 +138,11 @@ class ControllerChannelManager(controllerContext: ControllerContext, config: Kaf
 
   private def removeExistingBroker(brokerState: ControllerBrokerStateInfo) {
     try {
+      // shutdown the RequestSendThread before closing the NetworkClient to
+      // avoid the concurrent use of the non-threadsafe classes as described in KAFKA-4959
+      brokerState.requestSendThread.shutdown()
```

Comment:
Thanks for the review @ijuma.

Regarding:
> However, it's worth pointing out that this is still not safe.

Not sure if that's right.

By moving `brokerState.requestSendThread.shutdown()` infront of the `brokerState.networkClient.close()`, we actually have a synchronization barrier that hands off the NetworkClient from RequestSendThread to the ZkEventThread: namely, the call to `shutdownLatch.await()` in `ShutdownableThread.shutdown`. So we are guaranteed there is only a single thread accessing the NetworkClient.

Maybe I missed something?